pydantic/schema.py

```diff
@@ -196,7 +196,7 @@ def field_schema(
         known_models=known_models or set(),
     )
     # $ref will only be returned when there are no schema_overrides
-    if '$ref' in f_schema:
+    if '$ref' in f_schema or (field.allow_none and [x for x in f_schema.get('anyOf', []) if '$ref' in x]):
```

Comment:
can you explain what's going on here, perhaps include a comment, also this can be make more clear with 
```suggestion
    if '$ref' in f_schema or (field.allow_none and any('$ref' in x for x in f_schema.get('anyOf', []))):
```